File: 590.N-ary-Tree-Postorder-Traversal.py
# https://leetcode.com/problems/n-ary-tree-postorder-traversal/description/
#
# algorithms
# Easy (61.31%)
# Total Accepted:    8.4k
# Total Submissions: 13.7k
# beats 14.19% of python submissions


"""
# Definition for a Node.
class Node(object):
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""


class Solution(object):
    def postorder(self, root):
        """
        :type root: Node
        :rtype: List[int]
        """
        res = [[]]

        def resursive(node):
            if not node:
                return

            for child in node.children:
                resursive(child)

            res[0] += node.val,

        resursive(root)

        return res[0]

File: 590.N-ary-Tree-Postorder-Traversal.java
// https://leetcode.com/problems/n-ary-tree-postorder-traversal/
//
// algorithms
// Easy (69.84%)
// Total Accepted: 66,221
// Total Submissions: 94,815


/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
    public List<Integer> postorder(Node root) {
        ArrayList<Integer> l = new ArrayList<>();

        recursive(l, root);

        return l;
    }

    public void recursive(ArrayList l, Node node) {
        if (node == null) {
            return;
        }

        for (Node n : node.children) {
            recursive(l, n);
        }
        l.add(node.val);
    }
}